### PR TITLE
Remove avatar on user deletion

### LINKED ISSUE #6621: Profile picture survives deletion and recreation of account
### Steps to reproduce
1. Create account X and set a profile picture
2. Delete the account X
3. Create a new account X with the same username

### Expected behaviour
The profile picture is reset.

### Actual behaviour
The profile picture of the old account is visible

### Server configuration

**Operating system**: Fedora 26

**Web server:** Apache/2.4.27

**Database:** MariaDB/10.1.26-2.fc26

**PHP version:** 7.1.9-2.fc26

**Nextcloud version:** 12.0.3

**Updated from an older Nextcloud/ownCloud or fresh install:** Update from 12.0.2

**Where did you install Nextcloud from:** nextcloud.com zip file

**Signing status:**

**List of activated apps:**
Enabled:                                                                                                                                                         
  - activity: 2.5.2                      
  - apporder: 0.4.0                      
  - bruteforcesettings: 1.0.2            
  - calendar: 1.5.5                      
  - comments: 1.2.0                      
  - contacts: 1.5.3                      
  - dav: 1.3.0                           
  - federatedfilesharing: 1.2.0          
  - federation: 1.2.0
  - files: 1.7.2
  - files_downloadactivity: 1.1.1
  - files_markdown: 2.0.1
  - files_pdfviewer: 1.1.1
  - files_sharing: 1.4.0
  - files_texteditor: 2.4.1
  - files_trashbin: 1.2.0
  - files_versions: 1.5.0
  - files_videoplayer: 1.1.0
  - firstrunwizard: 2.1
  - gallery: 17.0.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - metadata: 0.5.0
  - nextcloud_announcements: 1.1
  - notes: 2.3.1
  - notifications: 2.0.0
  - oauth2: 1.0.5
  - password_policy: 1.2.2
  - previewgenerator: 1.0.6
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - sharebymail: 1.2.0
  - survey_client: 1.0.0
  - systemtags: 1.2.0
  - tasks: 0.9.5
  - theming: 1.3.0
  - twofactor_backupcodes: 1.1.1
  - twofactor_totp: 1.3.1
  - updatenotification: 1.2.0
  - workflowengine: 1.2.0

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #6621 

Basically I added a new event to the event dispatcher: `OCP\IUser::preDelete`
And then connected this to remove the avatar.

### CODE REVIEW

File: lib/private/Server.php [hunk 1]
```diff
@@ -1066,6 +1071,26 @@ public function __construct($webRoot, \OC\Config $config) {
...
+		$dispatcher->addListener('OCP\IUser::preDelete', function(GenericEvent $e) {
+			$manager = $this->getAvatarManager();
+			/** @var IUser $user */
+			$user = $e->getSubject();
+
+			try {
+				$avatar = $manager->getAvatar($user->getUID());
+				$avatar->remove();
+			} catch (\Exception $e) {
+				// Ignore exceptions
```
Comment (ChristophWurst) [inline lib/private/Server.php]: IMO would still make sense to log this, even if it's just level `debug` or `info` but if this code ever triggers errors nobody would notice and it makes this feature *useless*.

Comment (rullzer) [reply lib/private/Server.php]: Fair enough, let me do that.

Comment (rullzer) [reply lib/private/Server.php]: Done

**Review by ChristophWurst — APPROVED** [review]
Looks good!

File: lib/private/Server.php (hunk 1 above)
Comment (blizzz) [reply lib/private/Server.php]: could have removed the comment 🙊

File: lib/private/Server.php [hunk 2]
```diff
@@ -1084,6 +1089,28 @@ public function __construct($webRoot, \OC\Config $config) {
...
 		});
+
+		$this->connectDispatcher();
+	}
+
+	private function connectDispatcher() {
+		$dispatcher = $this->getEventDispatcher();
+
+		// Delete avatar on user deletion
+		$dispatcher->addListener('OCP\IUser::preDelete', function(GenericEvent $e) {
```
Comment (blizzz) [inline lib/private/Server.php]: Server container is not the proper place to put place this imho. Rather the AvatarManager.

File: lib/private/Server.php (hunk 1 above)
Comment (rullzer) [reply lib/private/Server.php]: We still kind of ignore it. We just also log it :P

File: lib/private/Server.php (hunk 2 above)
Comment (rullzer) [reply lib/private/Server.php]: I can't put it in the avatarmanager as that is only called once we need the avatarmanager.

Like if you delete a user via the provisioning API. The avatarManager is never initialised. And as such there is nothing listening in there as well.

Comment (blizzz) [reply lib/private/Server.php]: ah, rights, that's why all the other listeners are registered in base.php :)

Comment (ChristophWurst) [reply lib/private/Server.php]: FYI: a *service provider*, as the Laravel framework calls them, would be nice in this case.

https://laravel.com/docs/5.5/providers

Comment (rullzer) [reply lib/private/Server.php]: Yes :P But a bit out of scope. I moved it for this to the server as base.php is a long huge list of code and functions. IMO this has more of a place in Server.php to link everything together than in base.php

Comment (ChristophWurst) [reply lib/private/Server.php]: > Yes :P But a bit out of scope

I know. But well, maybe you feel highly motivated during a late night hacking session at the hackweek and want to integrate that :P

**Comment by blizzz** [comment]
tests are failing

**Comment by MorrisJobke** [comment]
@rullzer Ping

**Comment by rullzer** [comment]
@icewind1991 the errors happen in `UpdaterLegacyTest` because we try to access storage that isn't there. I'm not sure but it seems those tests to evil stuff. Can you have a look?

**Comment by MorrisJobke** [comment]
> @icewind1991 the errors happen in UpdaterLegacyTest because we try to access storage that isn't there. I'm not sure but it seems those tests to evil stuff. Can you have a look?

🏓

**Comment by MorrisJobke** [comment]
@icewind1991 Still failing: `PHP Fatal error:  Call to a member function getStorage() on null in /drone/src/github.com/nextcloud/server/lib/private/Files/View.php on line 1365`

**Comment by icewind1991** [comment]
all fixed

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Pull request merged**